libc/string_lenchr.c

```diff
@@ -148,6 +149,213 @@ TEST(string_len, big)
+	TEST_ASSERT_EQUAL_INT(sz, strspn((const char *)charSet, (const char *)charSet));
+}
+
+
+TEST(string_spn, big)
+{
+	char bigstr[PATH_MAX] = { 0 };
+	int sz;
+
+	/*
+	 * The length of the string is not restricted,
+	 * so we test one of the bigger values, which may be used
+	 * with remembering the last element must be a null term
+	 * zero to collect the data as the string from
+	 * declared space, that's why we memset one place less than its size
+	 */
+	memset(bigstr, 'a', sizeof(bigstr) - 1);
+
+	sz = sizeof(bigstr) - 1;
+	TEST_ASSERT_EQUAL_INT(sz, strcspn(bigstr, ""));
+	TEST_ASSERT_EQUAL_INT(sz, strspn(bigstr, bigstr));
+}
+
+
+TEST(string_spn, empty_output)
+{
+	char emptySet[BUFF_SIZE] = { 0 };
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn(emptySet, emptySet));
+	TEST_ASSERT_EQUAL_INT(0, strspn(emptySet, emptySet));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn(emptySet, ""));
+	TEST_ASSERT_EQUAL_INT(0, strspn(emptySet, ""));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn(emptySet, "abc"));
+	TEST_ASSERT_EQUAL_INT(0, strspn(emptySet, "abc"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("", "abc"));
+	TEST_ASSERT_EQUAL_INT(0, strspn("", "abc"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("", ""));
+	TEST_ASSERT_EQUAL_INT(0, strspn("", ""));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("\0abc", "abc"));
+	TEST_ASSERT_EQUAL_INT(0, strspn("\0abc", "abc"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("\0abc", ""));
+	TEST_ASSERT_EQUAL_INT(0, strspn("\0abc", ""));
+}
+
+
+TEST(string_spn, mixed_order)
+{
+
+	char *asciiSet = testdata_createCharStr(BUFF_SIZE),
+		 strHolder[BUFF_SIZE] = { 0 },
+		 supportSet[BUFF_SIZE] = { 0 };
+	int i;
+
+	asciiSet[BUFF_SIZE - 1] = 0;
+
+	/*
+	 * In those cases strcspn look for strings which not equal on the other hand strspn
+	 * looking for equal strings
+	 */
+	TEST_ASSERT_EQUAL_INT(0, strcspn("a", "a"));
+	TEST_ASSERT_EQUAL_INT(1, strspn("a", "a"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("abc", "a"));
+	TEST_ASSERT_EQUAL_INT(1, strspn("abc", "a"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("abc", "cba"));
+	TEST_ASSERT_EQUAL_INT(3, strspn("abc", "cba"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("abc", "abc1"));
+	TEST_ASSERT_EQUAL_INT(3, strspn("abc", "abc1"));
+
+	memset(strHolder, 'a', BUFF_SIZE - 3);
+	strHolder[BUFF_SIZE - 2] = 'b';
+	memset(supportSet, 'b', BUFF_SIZE - 3);
+	supportSet[BUFF_SIZE - 2] = 'a';
+
+	TEST_ASSERT_EQUAL_INT(BUFF_SIZE - 3, strcspn(strHolder, supportSet));
+	TEST_ASSERT_EQUAL_INT(0, strspn(strHolder, supportSet));
+
+	TEST_ASSERT_EQUAL_INT(BUFF_SIZE - 3, strcspn(supportSet, strHolder));
+	TEST_ASSERT_EQUAL_INT(0, strspn(supportSet, strHolder));
+
+	memcpy(supportSet, asciiSet, BUFF_SIZE - 1);
+
+	for (i = 0; i < BUFF_SIZE - 1; i++) {
+
+		/*
+		 * Checking if strcspn and strspn will search
+		 * the whole string before returning value and
+		 * he found an element of impact on the output
+		 */
+
+		/* Setting only the first element of set to values from asciiSet */
+		supportSet[0] = asciiSet[i];
```

Comment:
From what I see it's checking whether 2 same bytes are not a problem. You could check it, but it should be done in the separate test case and I would place there the strings like "aa", "aabbcc", "aaaaaa" as s2 and "abc" as s1 then. Not so many iterations and cases are needed in my opinion. I don't quite understand what you meant in the comment above, we could talk about it in the office.

Reply:
Did you add it somewhere?